### PR TITLE
Marathon Provider support

### PR DESCRIPTION
**Pull request by gokmen**
cc/ @rjeczalik

### CODE REVIEW

**Comment by gokmen** [comment]
@rjeczalik I'm done with this PR waiting for your input, fyi.

**Comment by rjeczalik** [comment]
@gokmen Assigning back. A number of changes, the most important one was to ship screen binary with klient. In order to have a single screen binary work across various linux distros I've compiled a static screen binary under Alpine Linux with musl replacement for glibc, with a small hack to workaround missing wtmp/utmp support in musl.

Since proxy does not work currently in development environment, I've tested webterm using my old evilkite client. I've also noticed that stack build modal sometimes losses events, especially when they are instant like in case of marathon.

Obligatory gif:

pic

**Comment by rjeczalik** [comment]
Can we also change default template to `cpus: 0.5` + `mem: 128`?

**Comment by rjeczalik** [comment]
@cihangir @ppknap Please review kloud- and klient-related code.

**Comment by rjeczalik** [comment]
Rebased, split vendor update out to a separate PR: https://github.com/koding/koding/pull/9565

**Review by ppknap — APPROVED** [review]
LGTM

**Comment by rjeczalik** [comment]
Rebased and added a test for unique labeling: https://github.com/koding/koding/pull/9547/commits/c84ba8890fac913646764a6967245b1aa0f9beb7

**Review by ppknap — APPROVED** [review]

File: go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden [hunk 1]
```diff
@@ -98,7 +111,7 @@
           "health_check": [
             {
               "command": {
-                "value": "curl -f -X GET http://$$HOST:$${PORT_56789}/kite"
+                "value": "true"
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden]: just a question, given the fact that i am not familiar with marathon,  why did we swicth from curl statement to true?

File: go/src/koding/kites/tunnelproxy/client.go [hunk 2]
```diff
@@ -135,8 +136,18 @@ func NewClient(opts *ClientOptions) (*Client, error) {
 		optsCopy.Log = logging.NewCustom("tunnelclient", optsCopy.Debug)
 	}
 
+	// TODO(rjeczalik): fix production to use WebSocket by default
```
Comment (cihangir) [inline go/src/koding/kites/tunnelproxy/client.go]: this is not viable for near future.

File: go/src/koding/klient/terminal/command.go [hunk 3]
```diff
@@ -135,7 +135,11 @@ func screenSessions(username string) []string {
 
 	// We need to use ls here, because /var/run/screen mount is only
 	// visible from inside of container. Errors are ignored.
-	out, _ := exec.Command("ls", "/var/run/screen/S-"+username).Output()
+	out, err := exec.Command("ls", "/var/run/screen/S-"+username).Output()
+	if err != nil {
```
Comment (cihangir) [inline go/src/koding/klient/terminal/command.go]: 👍

File: go/src/koding/klient/terminal/terminal.go [hunk 4]
```diff
@@ -253,7 +255,13 @@ func (t *Terminal) Connect(r *kite.Request) (interface{}, error) {
 	}
 
 	args = append(args, command.Args...)
-	cmd := exec.Command("/usr/bin/sudo", args...)
+	var cmd *exec.Cmd
+
+	if _, err := exec.LookPath("sudo"); err == nil {
```
Comment (cihangir) [inline go/src/koding/klient/terminal/terminal.go]: 1) you might cache the sudo path
2) you could use the return value from LookPath `exec.Command("/usr/bin/sudo", args...)`

File: go/src/koding/kites/tunnelproxy/client.go (hunk 2 above)
Comment (rjeczalik) [inline go/src/koding/kites/tunnelproxy/client.go]: Yes I know, it uses websocket for tunnelserver only. This TODO has no deadline :-)

File: go/src/koding/klient/terminal/terminal.go (hunk 4 above)
Comment (rjeczalik) [inline go/src/koding/klient/terminal/terminal.go]: I will change it to os.Stat. I think the path must always be `/usr/bin/sudo`, executing it from different path can be a security threat I believe.

File: go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden]: https://github.com/koding/koding/pull/9547/files#diff-12a6b9ca19c0ec2dc9242fd05655ce9eR26

The healtchecking is not needed to have working marathon provider - I wanted to be too correct to provide a healthcheck to ensure klient works. I need to both extend klient to allow for healthchecking (`curl .../kite` is not enough as it does not fail when klient has e.g. broken tunnel connection) and extend kloud provider for marathon with http healthchecks, but I'd need to map the ports for that.

File: go/src/koding/klient/terminal/terminal.go (hunk 4 above)
Comment (rjeczalik) [inline go/src/koding/klient/terminal/terminal.go]: Done.

File: go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden]: I've changed it to not inject dummy health checks for now.

**Comment by rjeczalik** [comment]
I've added support for multiline `cmd`, so it's possible to run ubuntu:14.04 like:

```
provider:
  marathon:
    url: "${var.marathon_url}"
    basic_auth_user: "${var.marathon_basic_auth_user}"
    basic_auth_password: "${var.marathon_basic_auth_password}"

resource:
  marathon_app:
    # this is the name of your app
    app:
      # 
      cmd: |-
        apt-get update
        apt-get install -y ca-certificates
        /opt/kite/klient/klient
      
      container:
        docker:
        - image: ubuntu:14.04
          network: BRIDGE

      cpus: 1.0
      mem: 256
```

It does not work out of the box due to https://github.com/docker/docker/issues/3825.

**Comment by rjeczalik** [comment]
Added one more change that injects ca-certificates.crt into container if it does not exist, ptal.

**Comment by rjeczalik** [comment]
@cihangir @gokmen ping

**Comment by gokmen** [comment]
LGTM but I'm unable to approve it 

image

**Review by rjeczalik — APPROVED** [review]

**Comment by rjeczalik** [comment]
@gokmen Did it for you! 💥

**Pull request merged**